torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -596,6 +596,116 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     std::string traceKeyEnd_;
   };
 
+  // Class that runs as part of a separate thread aside from watchdog
+  // thread because we need to check the heartbeat from watchdog thread
+  // so that when we get stuck in some NCCL/CUDA calls,
+  // we can dump the debugging information and abort the process./
+  // The thread is per class not per instance.
+  class HeartbeatMonitor {
+   public:
+    // Return the singleton instance of HeartbeatMonitor.
+    static HeartbeatMonitor* get() {
+      static HeartbeatMonitor instance;
```

Comment:
nit: should put static in a .cpp file otherwise you could end up with multiple if you reference this function from different CPP files. I assume that's not intentional?

Reply:
oh I see.. I can move it to cpp file.